### PR TITLE
make boost::filesystem optional

### PR DESCRIPTION
**Pull request by eagleoflqj**
Intent: on macOS enable tools but exclude migrators; on iOS disable them all.

### CODE REVIEW

File: tools/CMakeLists.txt [hunk 1]
```diff
@@ -24,6 +24,7 @@ target_link_libraries(libime_tabledict LibIME::Table)
 install(TARGETS libime_tabledict DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT tools)
 add_executable(LibIME::tabledict ALIAS libime_tabledict)
 
+if (NOT APPLE)
 add_executable(libime_migrate_fcitx4_table libime_migrate_fcitx4_table.cpp)
```
Comment (lotem) [inline tools/CMakeLists.txt:28]: 這是什麼？怎麼來的。我咋不知道呢。

**Review by lotem — APPROVED** [review]

File: tools/CMakeLists.txt (hunk 1 above)
Comment (eagleoflqj) [reply tools/CMakeLists.txt:28]: 你串台了吧 🌚 咋发现这个PR的

Comment (lotem) [reply tools/CMakeLists.txt:28]: 不懂。反正無腦幫你+1.

**Comment by wengxt** [comment]
boost filesystem is suppose to be optional if std::filesystem is available. 

There might be some issue on the __GNUC__ <= 8 condition when build against clang, can you help fix that instead?

basically, libstdc++ in gcc 7 doesn't have usable std::filesystem

**Comment by eagleoflqj** [comment]
Wouldn't it be preferable to remove migration tools? After all fcitx4 was archived, and old versions of libime can always do the one-time job.

**Comment by wengxt** [comment]
> Wouldn't it be preferable to remove migration tools? After all fcitx4 was archived, and old versions of libime can always do the one-time job.

Well, some people may have random binary files available on internet so it could still be useful. I have no plan to remove them anytime soon.

**Comment by wengxt** [comment]
If you need an example on std::filesystem detection, here's something I found on internet https://github.com/ccache/ccache/pull/1190/files

**Comment by wengxt** [comment]
Basically what in my mind is, run a CheckCXXSourceCompiles against the API that we use (there's only one), if fail, then fallback to boost filesytem. I believe whatever you care should have usable std::filesystem

File: CMakeLists.txt [hunk 2]
```diff
@@ -34,7 +35,24 @@ pkg_check_modules(ZSTD REQUIRED IMPORTED_TARGET "libzstd")
 find_package(Fcitx5Utils REQUIRED)
 include("${FCITX_INSTALL_CMAKECONFIG_DIR}/Fcitx5Utils/Fcitx5CompilerSettings.cmake")
 
-find_package(Boost 1.61 REQUIRED COMPONENTS iostreams filesystem)
+include(CheckCXXSourceCompiles)
+check_cxx_source_compiles("
+#include <filesystem>
+int main() {
+    return std::filesystem::is_regular_file(\"/\") ? 0 : 1;
```
Comment (wengxt) [inline CMakeLists.txt:42]: Can you use std::filesystem::absolute to test?

File: tools/CMakeLists.txt [hunk 3]
```diff
@@ -24,12 +24,14 @@ target_link_libraries(libime_tabledict LibIME::Table)
 install(TARGETS libime_tabledict DESTINATION ${CMAKE_INSTALL_BINDIR} COMPONENT tools)
 add_executable(LibIME::tabledict ALIAS libime_tabledict)
 
+if (NOT APPLE)
```
Comment (wengxt) [inline tools/CMakeLists.txt:27]: Remove this

File: CMakeLists.txt [hunk 4]
```diff
@@ -34,7 +35,24 @@ pkg_check_modules(ZSTD REQUIRED IMPORTED_TARGET "libzstd")
...
+}
+" HAS_STD_FILESYSTEM)
+
+if (HAS_STD_FILESYSTEM)
```
Comment (wengxt) [inline CMakeLists.txt:46]: I would rewrite it as
find_package(Boost ... iostreams)
if (NOT HAS_STD_FILESYSTEM)
      find_package(Boost REQUIRED COMPONENTS filesystem)
endif()

in tools/CMakeLists.txt

if (NOT HAS_STD_FILESYSTEM)
    target_compile_definitiones(.... -D....)
    target_link_libraries(.... )
endif()


target_link_libraries can be repeated mulitple times to add more deps.
and avoid use add_definitions on unrelated targets.

**Pull request merged**